File: killstreek/src/main/java/dev/kkorolyov/killstreek/item/Item.java
package dev.kkorolyov.killstreek.item;

import dev.kkorolyov.killstreek.media.Sprite;
import dev.kkorolyov.pancake.platform.action.Action;
import dev.kkorolyov.pancake.platform.entity.Entity;
import dev.kkorolyov.simplestructs.WeightedDistribution;

import java.util.NoSuchElementException;

/**
 * Applies effects to entities.
 */
public abstract class Item {
	/** An effect which does nothing, used for padding the pool of random effects */
	public static final Action NOOP_EFFECT = entity -> {};
	private static int idCounter;

	private static int generateId() {
		return idCounter++;
	}

	private final int id;
	private final String name;
	private final Sprite sprite;
	private final WeightedDistribution<Action> effects = new WeightedDistribution<>();

	/**
	 * Constructs a new item with a unique ID.
	 * @param name item name
	 * @param sprite item visual
	 */
	public Item(String name, Sprite sprite) {
		this.id = generateId();
		this.name = name;
		this.sprite = sprite;
	}

	/**
	 * Applies a randomly-select effect from this item's effect pool to an entity.
	 * @param entity entity receiving effect
	 * @throws NoSuchElementException if this item has no effects
	 */
	public void apply(Entity entity) {
		effects.get().apply(entity);
	}

	/**
	 * Adds an effect with weight 1 to this item.
	 * @param effect effect on entity
	 * @return {@code this}
	 */
	public Item addEffect(Action effect) {
		return addEffect(effect, 1);
	}
	/**
	 * Adds an effect to this item.
	 * @param effect effect on entity
	 * @param weight effect frequency in relation to all other effects of this item
	 * @return {@code this}
	 */
	public Item addEffect(Action effect, int weight) {
		effects.add(effect, weight);
		return this;
	}

	/** @return unique item ID */
	public int getId() {
		return id;
	}

	/** @return item name */
	public String getName() {
		return name;
	}

	/** @return item sprite */
	public Sprite getSprite() {
		return sprite;
	}

	/** @return maximum number of items of this type in a single stack */
	public abstract int getMaxStackSize();
}

File: killstreek/src/main/kotlin/dev/kkorolyov/killstreek/item/Item.kt
package dev.kkorolyov.killstreek.item

import dev.kkorolyov.killstreek.media.Sprite
import dev.kkorolyov.pancake.platform.action.Action
import dev.kkorolyov.pancake.platform.entity.Entity
import dev.kkorolyov.simplestructs.WeightedDistribution

/** An effect which does nothing, used for padding the pool of random effects */
val NOOP_EFFECT: Action = Action {}

private var idCounter = 0

/**
 * Applies effects to entities.
 * @constructor constructs a new item with a unique ID
 */
abstract class Item(
		/** item name */
		val name: String,
		/** item visual */
		val sprite: Sprite
) : Action {
	/** unique item ID */
	val id = idCounter++
	private val effects: WeightedDistribution<Action> = WeightedDistribution()

	/** maximum number of items of this type that may exist in a single stack */
	abstract val maxStackSize: Int

	/**
	 * Applies a randomly-selected effect from this item's effect pool to [entity].
	 */
	override fun apply(entity: Entity) {
		effects.get().apply(entity)
	}

	/**
	 * Adds [effect] with selection [weight] to this item and returns this item.
	 */
	fun addEffect(effect: Action, weight: Int): Item {
		effects.add(effect, weight)
		return this
	}
}
